### ISSUE TITLE
Simple way to manage and release GPU memory in colab

### ISSUE THREAD

**Issue by HarikrishnanBalagopal** [tensorflow/tensorflow#31935]
<em>Please make sure that this is a bug. As per our [GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md), we only address code/doc bugs, performance issues, feature requests and build/installation issues on GitHub. tag:bug_template</em>

**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): Yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Google Colab Ubuntu 18.04.2 LTS (Bionic Beaver)
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device: NA
- TensorFlow installed from (source or binary): Google Colab has tensorflow preinstalled
- TensorFlow version (use command below): tensorflow-gpu 1.14.0
- Python version: 3
- Bazel version (if compiling from source): NA
- GCC/Compiler version (if compiling from source): NA
- CUDA/cuDNN version: Cuda compilation tools, release 10.0, V10.0.130
- GPU model and memory: Google Colab GPU Tesla T4, Memory: 15079MiB

You can collect some of this information using our environment capture
[script](https://github.com/tensorflow/tensorflow/tree/master/tools/tf_env_collect.sh)
You can also obtain the TensorFlow version with: 1. TF 1.0: `python -c "import
tensorflow as tf; print(tf.GIT_VERSION, tf.VERSION)"` 2. TF 2.0: `python -c
"import tensorflow as tf; print(tf.version.GIT_VERSION, tf.version.VERSION)"`

**Describe the current behavior**
For my research work I need to build lots of different convolutional GAN models and train them.
For this I build temporary models inside functions and test them. Once the function is done executing the models are no longer needed.
A simple example:
```python
def test_model():
   model = build_model()
   inputs = ...
   outputs = model(inputs)
   with tf.Session() as sess:
      sess.run(tf.global_variables_initializer())
      results = sess.run(outputs)
   # evaluate results

test_model() # call the test function, once this is over we never use the model again.
```
However currently the models persist (can be seen in %tensorboard) and continue to fill up
GPU memory. Eventually I start getting warning about GPU memory usage and start getting OOM errors. At this point I can't build new models or train any existing ones.

I have already tried lots of different suggestions on how to release GPU memory
https://github.com/tensorflow/tensorflow/issues/1578
https://github.com/tensorflow/tensorflow/issues/19731
https://github.com/tensorflow/tensorflow/issues/17048
and several stackoverflow suggestions to no effect.

This problem is specific to a Jupter notebook based workflow (such as on Google Colab).
A workflow that uses python files will not encounter this issue since all the GPU memory is released automatically once the python interpreter finishes.

**Describe the expected behavior**
- Please implement or suggest a way to release GPU memory being used by unneeded models in Google Colab/Jupter notebooks.
- It would be nice to be able to release memory being used by specific models (that are no longer necessary) rather than resetting the runtime every time I run out of memory (which is often).
- Some way to build models so that the GPU memory they occupy gets automatically released when they go out of scope would also be appreciated.

**Code to reproduce the issue**
Provide a reproducible test case that is the bare minimum necessary to generate the problem.

**Other info / logs**
Include any logs or source code that would be helpful to diagnose the problem. If including tracebacks, please include the full traceback. Large logs and files should be attached.

**Comment by jmwoloso** [comment]
@HarikrishnanBalagopal I was running 10 training loops on colab and running into OOM errors after the 8th iteration. Putting the following at the end of the loop helped.
`tf.reset_default_graph()`

**Comment by jvishnuvardhan** [comment]
@HarikrishnanBalagopal Did you try @jmwoloso Suggestion. In some cases, you could also use `tf.keras.backend.clear_session()` to release the session memory. Thanks!

**Comment by HarikrishnanBalagopal** [comment]
I am currently calling ` tf.reset_default_graph()` if I ever pollute the default graph and changed my code to:
```python
def test_model():
   g_1 = tf.Graph()
   with g_1.as_default():
      model = build_model()
      inputs = ...
      outputs = model(inputs)
   with tf.Session() as sess:
      sess.run(tf.global_variables_initializer())
      results = sess.run(outputs)
      # evaluate results

test_model() # call the test function, once this is over we never use the model again.
```
which seems to work fine so far. The graph gets garbage collected and haven't had any OOM errors yet, so I am closing the issue.

**Comment by bsurya27** [comment]
Is there a way where I can clear specific things in the gpu? I have a few variables which I do not want to erase and there are others which I want to.
